casanovo/denovo/model.py

```diff
@@ -171,16 +178,613 @@ def forward(
+
+        return peptides, aa_scores
+
+    def beam_search_decode(
+        self, spectra: torch.Tensor, precursors: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Beam search decode the spectra.
+
+        Return the highest scoring peptide, within the precursor m/z tolerance
+        whenever possible.
+
+        Parameters
+        ----------
+        spectra : torch.Tensor of shape (n_spectra, n_peaks, 2)
+            The spectra for which to predict peptide sequences.
+            Axis 0 represents an MS/MS spectrum, axis 1 contains the peaks in
+            the MS/MS spectrum, and axis 2 is essentially a 2-tuple specifying
+            the m/z-intensity pair for each peak. These should be zero-padded,
+            such that all of the spectra in the batch are the same length.
+        precursors : torch.Tensor of size (n_spectra, 3)
+            The measured precursor mass (axis 0), precursor charge (axis 1),
+            and precursor m/z (axis 2) of each MS/MS spectrum.
+
+        Returns
+        -------
+        scores : torch.Tensor of shape (n_spectra, max_length, n_amino_acids)
+            The score for each amino acid.
+        tokens : torch.Tensor of shape (n_spectra, max_length)
+            The token sequence for each spectrum.
+        """
+        memories, mem_masks = self.encoder(spectra)
+
+        # Sizes:
+        batch = spectra.shape[0]  # B
+        length = self.max_length + 1  # L
+        vocab = self.decoder.vocab_size + 1  # V
+        beam = self.n_beams  # S
+
+        # Initialize scores and tokens:
+        scores = torch.full(
+            size=(batch, length, vocab, beam), fill_value=torch.nan
+        )
+        scores = scores.type_as(spectra)
+        tokens = torch.zeros(batch, length, beam, dtype=torch.int64).to(
+            self.encoder.device
+        )
+        # Keep track if terminated beams have fitting precursor m/z
+        is_beam_prec_fit = (batch * beam) * [False]
+
+        # Create cache for decoded beams
+        (
+            cache_scores,
+            cache_tokens,
+            cache_next_idx,
+            cache_pred_seq,
+            cache_pred_score,
+        ) = self._create_beamsearch_cache(scores, tokens)
+
+        # Create the first prediction
+        pred, _ = self.decoder(None, precursors, memories, mem_masks)
+        _, idx = torch.topk(pred[:, 0, :], beam, dim=1)
+        tokens[:, 0, :] = idx
+        scores[:, :1, :, :] = einops.repeat(pred, "B L V -> B L V S", S=beam)
+
+        # Make precursors and memories the right shape for decoding:
+        precursors = einops.repeat(precursors, "B L -> (B S) L", S=beam)
+        mem_masks = einops.repeat(mem_masks, "B L -> (B S) L", S=beam)
+        memories = einops.repeat(memories, "B L V -> (B S) L V", S=beam)
+
+        # The main decoding loop:
+        for idx in range(1, self.max_length + 1):
+            scores = einops.rearrange(scores, "B L V S -> (B S) L V")
```

Comment:
**question:** Can this be moved outside of the loop?

Reply:
Moved it out of the loop and added reshaping inside `_get_topk_beams `